File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/02．战斗区域/04．圆形安全区组.ts
/** @noSelfInFile */

import type { 机制清理篮子 } from "../06．机制清理/01．机制清理篮子";
import { 创建技能提示圈 } from "../../02．通用函数/16．技能提示圈工厂";

const jass = require("jass.common") as any;

const GetUnitX = jass.GetUnitX as (u: any) => number;
const GetUnitY = jass.GetUnitY as (u: any) => number;

export interface 圆形安全区配置 {
  ID: string;
  X: number;
  Y: number;
  半径: number;
  名称?: string;
}

export interface 圆形安全区组参数 {
  清理?: 机制清理篮子;
  名称: string;
  安全区列表: 圆形安全区配置[];
  默认显示提示?: boolean;
  提示持续秒?: number;
}

export interface 圆形安全区组 {
  readonly 名称: string;
  取列表(): 圆形安全区配置[];
  取安全区(ID: string): 圆形安全区配置 | undefined;
  点是否安全(x: number, y: number): boolean;
  单位是否安全(单位: any): boolean;
  显示提示(持续秒?: number): void;
  销毁(): void;
}

function 距离平方(this: void, x1: number, y1: number, x2: number, y2: number): number {
  const dx = x1 - x2;
  const dy = y1 - y2;
  return dx * dx + dy * dy;
}

class 圆形安全区组实现 implements 圆形安全区组 {
  readonly 名称: string;
  private 参数: 圆形安全区组参数;
  private 已销毁 = false;

  constructor(参数: 圆形安全区组参数) {
    this.名称 = 参数.名称;
    this.参数 = 参数;
    if (参数.默认显示提示) this.显示提示(参数.提示持续秒);
  }

  取列表(): 圆形安全区配置[] {
    return this.参数.安全区列表;
  }

  取安全区(ID: string): 圆形安全区配置 | undefined {
    const 列表 = this.参数.安全区列表;
    for (let i = 0; i < 列表.length; i++) {
      if (列表[i].ID === ID) return 列表[i];
    }
    return undefined;
  }

  点是否安全(x: number, y: number): boolean {
    const 列表 = this.参数.安全区列表;
    for (let i = 0; i < 列表.length; i++) {
      const 区 = 列表[i];
      if (距离平方(x, y, 区.X, 区.Y) <= 区.半径 * 区.半径) return true;
    }
    return false;
  }

  单位是否安全(单位: any): boolean {
    if (单位 == null || 单位 === 0) return false;
    return this.点是否安全(GetUnitX(单位), GetUnitY(单位));
  }

  显示提示(持续秒?: number): void {
    const 列表 = this.参数.安全区列表;
    for (let i = 0; i < 列表.length; i++) {
      const 区 = 列表[i];
      创建技能提示圈({
        类型: "白色安全圆",
        X: 区.X,
        Y: 区.Y,
        半径: 区.半径,
        持续时间: 持续秒 ?? this.参数.提示持续秒 ?? 3,
      });
    }
  }

  销毁(): void {
    this.已销毁 = true;
  }
}

export function 创建圆形安全区组(this: void, 参数: 圆形安全区组参数): 圆形安全区组 {
  const 实例 = new 圆形安全区组实现(参数);
  if (参数.清理 != null) {
    参数.清理.登记清理(参数.名称, function 圆形安全区组清理(this: void): void {
      实例.销毁();
    });
  }
  return 实例;
}

File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/02．战斗区域/04．圆形安全区组.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____16_FF0E_6280_80FD_63D0_793A_5708_5DE5_5382 = require("系统.03．技能系统.00．技能模板+函数.02．通用函数.16．技能提示圈工厂")
local _____521B_5EFA_6280_80FD_63D0_793A_5708 = ____16_FF0E_6280_80FD_63D0_793A_5708_5DE5_5382["创建技能提示圈"]
local jass = require("jass.common")
local GetUnitX = jass.GetUnitX
local GetUnitY = jass.GetUnitY
local function _____8DDD_79BB_5E73_65B9(x1, y1, x2, y2)
    local dx = x1 - x2
    local dy = y1 - y2
    return dx * dx + dy * dy
end
local _____5706_5F62_5B89_5168_533A_7EC4_5B9E_73B0 = __TS__Class()
_____5706_5F62_5B89_5168_533A_7EC4_5B9E_73B0.name = "圆形安全区组实现"
function _____5706_5F62_5B89_5168_533A_7EC4_5B9E_73B0.prototype.____constructor(self, _____53C2_6570)
    self["已销毁"] = false
    self["名称"] = _____53C2_6570["名称"]
    self["参数"] = _____53C2_6570
    if _____53C2_6570["默认显示提示"] then
        self["显示提示"](self, _____53C2_6570["提示持续秒"])
    end
end
_____5706_5F62_5B89_5168_533A_7EC4_5B9E_73B0.prototype["取列表"] = function(self)
    return self["参数"]["安全区列表"]
end
_____5706_5F62_5B89_5168_533A_7EC4_5B9E_73B0.prototype["取安全区"] = function(self, ID)
    local _____5217_8868 = self["参数"]["安全区列表"]
    do
        local i = 0
        while i < #_____5217_8868 do
            if _____5217_8868[i + 1].ID == ID then
                return _____5217_8868[i + 1]
            end
            i = i + 1
        end
    end
    return nil
end
_____5706_5F62_5B89_5168_533A_7EC4_5B9E_73B0.prototype["点是否安全"] = function(self, x, y)
    local _____5217_8868 = self["参数"]["安全区列表"]
    do
        local i = 0
        while i < #_____5217_8868 do
            local _____533A = _____5217_8868[i + 1]
            if _____8DDD_79BB_5E73_65B9(x, y, _____533A.X, _____533A.Y) <= _____533A["半径"] * _____533A["半径"] then
                return true
            end
            i = i + 1
        end
    end
    return false
end
_____5706_5F62_5B89_5168_533A_7EC4_5B9E_73B0.prototype["单位是否安全"] = function(self, _____5355_4F4D)
    if _____5355_4F4D == nil or _____5355_4F4D == 0 then
        return false
    end
    return self["点是否安全"](
        self,
        GetUnitX(_____5355_4F4D),
        GetUnitY(_____5355_4F4D)
    )
end
_____5706_5F62_5B89_5168_533A_7EC4_5B9E_73B0.prototype["显示提示"] = function(self, _____6301_7EED_79D2)
    local _____5217_8868 = self["参数"]["安全区列表"]
    do
        local i = 0
        while i < #_____5217_8868 do
            local _____533A = _____5217_8868[i + 1]
            _____521B_5EFA_6280_80FD_63D0_793A_5708({
                ["类型"] = "白色安全圆",
                X = _____533A.X,
                Y = _____533A.Y,
                ["半径"] = _____533A["半径"],
                ["持续时间"] = _____6301_7EED_79D2 or self["参数"]["提示持续秒"] or 3
            })
            i = i + 1
        end
    end
end
_____5706_5F62_5B89_5168_533A_7EC4_5B9E_73B0.prototype["销毁"] = function(self)
    self["已销毁"] = true
end
____exports["创建圆形安全区组"] = function(_____53C2_6570)
    local _____5B9E_4F8B = __TS__New(_____5706_5F62_5B89_5168_533A_7EC4_5B9E_73B0, _____53C2_6570)
    if _____53C2_6570["清理"] ~= nil then
        local ____self_0 = _____53C2_6570["清理"]
        ____self_0["登记清理"](
            ____self_0,
            _____53C2_6570["名称"],
            function()
                _____5B9E_4F8B["销毁"](_____5B9E_4F8B)
            end
        )
    end
    return _____5B9E_4F8B
end
return ____exports
